src/librustc_resolve/lib.rs

```diff
@@ -3457,6 +3536,123 @@ impl<'a, 'tcx> Resolver<'a, 'tcx> {
         found_traits
     }
 
+    /// Looks up a candidate entities with a given name and filters them using
+    /// the supplied predicate. The lookup spans across all crates. This method
+    /// should be used to generate suggestions when name resolution fails.
+    ///
+    /// NOTE: We are not looking into imports, so we will ignore all private-
+    /// made-public entities through patterns like this:
+    /// ```
+    /// pub mod foo {
+    ///     mod bar { trait T {}; };
+    ///     // make private T accessible through an import:
+    ///     pub use bar::T as T;
+    /// }
+    fn lookup_candidates<FilterFn>(&mut self,
+                                   lookup_name: Name,
+                                   namespace: Namespace,
+                                   filter_fn: FilterFn) -> SuggestedCandidates
+        where FilterFn: Fn(Def) -> bool {
+
+        let mut lookup_results_accessible = Vec::new();
+        let mut lookup_results_everything = Vec::new();
+        let mut lookup_modules = Vec::new();
+        lookup_modules.push((self.graph_root, Vec::new(), true));
+
+        while let Some((in_module, path_segments, module_is_public)) = lookup_modules.pop() {
```

Comment:
nit: s/`module_is_public`/`module_is_accessible` -- `is_public` means this module was defined with `pub`